### PR TITLE
[#176678865] Rename revocation endpoint and method

### PR DESCRIPTION
**Pull request by AleDore**
#### List of Changes
- Rename revocation endpoint and method to status upsert

#### Motivation and Context
In order to reach better API definitions we should rename revocation endpoint. The purpose is to map a revocation as a status upsert, that allows to use a body with the revocation reason.

Please refer to related [story](https://www.pivotaltracker.com/story/show/176678865) for further informations.

#### How Has This Been Tested?
It has been tested by performing unit tests.
No tests needed

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: RevokeCgn/function.json [hunk 1]
```diff
@@ -5,9 +5,9 @@
       "type": "httpTrigger",
       "direction": "in",
       "name": "req",
-      "route": "api/v1/cgn/{fiscalcode}",
+      "route": "api/v1/cgn/{fiscalcode}/revocation",
```
Comment (gunzip) [inline RevokeCgn/function.json:8]: so we'll get a neww endpoint to get revocations?
ie. GET api/v1/cgn/{fiscalcode}/revocation/{id} ?

Comment (AleDore) [reply RevokeCgn/function.json:8]: I got your expectation but I'd rather avoid to write multiple endpoint to reach the same resource, namely `userCgn` that is served by `/cgn/status/{fiscalcode}`. There's another 2 options:
- Rename the revocation endpoint to `api/v1/cgn/{fiscalcode}` with PUT method with no impacts on the handler's code
- Rewrite the `RevokeCgn` as `UpsertCgn` that allows to upsert `userCgn` with all the possible statuses.

IMHO the first one is better so we can keep activation and revocation separated. What do you think @gunzip ?

**Comment by gunzip** [comment]
> Rename the revocation endpoint to api/v1/cgn/{fiscalcode} with PUT method and no impacts on the handler's code

what I meant is that if you're doing a put / post then you're supposed to provide a way to retrieve the resource (get) at least to check the status (otherwise it's fire & forget). I suspect that doing a patch / put (upsert) is the right way here since you're going to check the status doing a get on api/v1/cgn/{fiscalcode}. another option is to have api/v1/cgn/{fiscalcode}/status (if you want to avoid a full upsert) and put there

File: UpsertCgnStatus/function.json [hunk 2]
```diff
@@ -5,9 +5,9 @@
       "type": "httpTrigger",
       "direction": "in",
       "name": "req",
-      "route": "api/v1/cgn/{fiscalcode}",
+      "route": "api/v1/cgn/status/{fiscalcode}",
```
Comment (gunzip) [inline UpsertCgnStatus/function.json:8]:
```suggestion
      "route": "api/v1/cgn/{fiscalcode}/status",
```

Comment (BurnedMarshal) [inline UpsertCgnStatus/function.json:8]: `status` is a property of resource `cgn/{fiscalcode}`, IMHO the right URL should be `api/v1/cgn/{fiscalcode}/status`

File: UpsertCgnStatus/handler.ts [hunk 3]
```diff
@@ -38,55 +48,72 @@ import { checkUpdateCgnIsRunning } from "../utils/orchestrators";
...
-export function RevokeCgnHandler(
+const toCgnStatus = (cgnStatusUpsertRequest: CgnStatusUpsertRequest) => {
+  return cgnStatusUpsertRequest.action === ActionEnum.REVOKE
+    ? {
+        revocation_date: new Date(),
+        revocation_reason: cgnStatusUpsertRequest.revocation_reason,
+        status: StatusEnum.REVOKED
+      }
+    : {
+        status: PendingStatusEnum.PENDING
```
Comment (gunzip) [inline UpsertCgnStatus/handler.ts:72]: may you add a comment here on what's the meaning of 'pending' and on why every other request causes a pending status?

File: UpsertCgnStatus/handler.ts [hunk 4]
```diff
@@ -38,55 +48,72 @@ import { checkUpdateCgnIsRunning } from "../utils/orchestrators";
...
+      };
+};
+
+export function UpsertCgnStatusHandler(
   userCgnModel: UserCgnModel,
-  logPrefix: string = "RevokeCgnHandler"
-): IRevokeCgnHandler {
-  return async (context, fiscalCode, revocationReq) => {
+  logPrefix: string = "UpsertCgnHandler"
```
Comment (gunzip) [inline UpsertCgnStatus/handler.ts:78]:
```suggestion
  logPrefix: string = "UpsertCgnStatusHandler"
```

File: UpsertCgnStatus/index.ts [hunk 5]
```diff
@@ -36,7 +36,7 @@ const app = express();
 secureExpressApp(app);
 
 // Add express route
-app.delete("/api/v1/cgn/:fiscalcode", RevokeCgn(userCgnModel));
+app.post("/api/v1/cgn/status/:fiscalcode", UpsertCgnStatus(userCgnModel));
```
Comment (gunzip) [inline UpsertCgnStatus/index.ts:39]:
```suggestion
app.post("/api/v1/cgn/:fiscalcode/status", UpsertCgnStatus(userCgnModel));

```

File: openapi/index.yaml [hunk 6]
```diff
@@ -174,14 +175,33 @@ definitions:
...
+    required:
+      - action
+  CgnStatusRevocationRequest:
+    type: object
+    properties:
+      action:
+        type: string
+        x-extensible-enum:
+          - REVOKE
       revocation_reason:
```
Comment (BurnedMarshal) [inline openapi/index.yaml:194]: Can we rename this prop with a more generic `reason`? This could be used for all status update requests

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: openapi/index.yaml (hunk 6 above)
Comment (AleDore) [reply openapi/index.yaml:194]: Nope the `reason` attribute is specifically defined for `REVOKED` CGNs. :)

File: UpsertCgnStatus/function.json (hunk 2 above)
Comment (AleDore) [reply UpsertCgnStatus/function.json]: Fixed in [0f2f273](https://github.com/pagopa/io-functions-cgn/pull/12/commits/0f2f273bd086ff2a263a5e50104cb22cbcb96cc4)

File: UpsertCgnStatus/index.ts (hunk 5 above)
Comment (AleDore) [reply UpsertCgnStatus/index.ts]: Fixed in [0f2f273](https://github.com/pagopa/io-functions-cgn/pull/12/commits/0f2f273bd086ff2a263a5e50104cb22cbcb96cc4)

File: UpsertCgnStatus/handler.ts (hunk 4 above)
Comment (AleDore) [reply UpsertCgnStatus/handler.ts]: Fixed in [0f2f273](https://github.com/pagopa/io-functions-cgn/pull/12/commits/0f2f273bd086ff2a263a5e50104cb22cbcb96cc4)

File: UpsertCgnStatus/handler.ts (hunk 3 above)
Comment (AleDore) [reply UpsertCgnStatus/handler.ts:76]: Added in [0f2f273](https://github.com/pagopa/io-functions-cgn/pull/12/commits/0f2f273bd086ff2a263a5e50104cb22cbcb96cc4)

File: UpsertCgnStatus/function.json (hunk 2 above)
Comment (AleDore) [reply UpsertCgnStatus/function.json]: Fixed in [0f2f273](https://github.com/pagopa/io-functions-cgn/pull/12/commits/0f2f273bd086ff2a263a5e50104cb22cbcb96cc4)

**Review by gunzip — APPROVED** [review]

**Pull request merged**